### ISSUE TITLE
wrong output of salt-api when working with salt-syndic

### ISSUE THREAD

**Issue by vinian** [saltstack/salt#48000]
### Description of Issue/Question

when firing salt execute function by salt-api, even if the function is executed on minions behind salt-syndic, but the output show "minion did not return. [No response]"

execute test on master of master
```
vagrant@master:~$ bash cherry-t.sh
+ curl http://localhost:8000/login -c ./cookies.txt -H 'Accept: application/x-yaml' -d username=saltdev -d password=saltdev -d eauth=pam
return:
- eauth: pam
  expire: 1528336871.5202
  perms:
  - '*':
    - test.*
    - state.*
    - grains.*
    - pkg.*
  start: 1528293671.520196
  token: 8ec97d066c7aa6fd104c70fcb04c523231d13dc6
  user: saltdev
+ curl http://localhost:8000 -b ./cookies.txt -H 'Accept: application/x-yaml' -d client=local -d tgt=minion01,minion02,master-syndic -d fun=state.sls -d tgt_type=list -d arg=test
return:
- master-syndic:
    cmd_|-test cmd_|-uptime_|-run:
      __id__: test cmd
      __run_num__: 0
      __sls__: test
      changes:
        pid: 24026
        retcode: 0
        stderr: ''
        stdout: ' 14:01:12 up  6:28,  1 user,  load average: 0.07, 0.03, 0.00'
      comment: Command "uptime" run
      duration: 20.197
      name: uptime
      result: true
      start_time: '14:01:12.172091'
  minion01: Minion did not return. [No response]
  minion02: Minion did not return. [No response]
```

event on master of mater
```
salt/job/20180606140111583476/new       {
    "_stamp": "2018-06-06T14:01:11.586445",
    "arg": [
        "test"
    ],
    "fun": "state.sls",
    "jid": "20180606140111583476",
    "minions": [
        "master-syndic"
    ],
    "missing": [
        "minion01",
        "minion02"
    ],
    "tgt": "minion01,minion02,master-syndic",
    "tgt_type": "list",
    "user": "saltdev"
}
20180606140111583476    {
    "_stamp": "2018-06-06T14:01:12.103829",
    "minions": [
        "minion01",
        "minion02"
    ]
}
syndic/master-syndic/20180606140111583476       {
    "_stamp": "2018-06-06T14:01:12.113088",
    "minions": [
        "minion01",
        "minion02"
    ]
}
salt/job/20180606140111583476/new       {
    "_stamp": "2018-06-06T14:01:12.120565",
    "arg": [
        "test"
    ],
    "fun": "state.sls",
    "jid": "20180606140111583476",
    "minions": [
        "minion01",
        "minion02"
    ],
    "missing": [
        "master-syndic"
    ],
    "tgt": "minion01,minion02,master-syndic",
    "tgt_type": "list",
    "user": "saltdev"
}
syndic/master-syndic/salt/job/20180606140111583476/new  {
    "_stamp": "2018-06-06T14:01:12.131604",
    "arg": [
        "test"
    ],
    "fun": "state.sls",
    "jid": "20180606140111583476",
    "minions": [
        "minion01",
        "minion02"
    ],
    "missing": [
        "master-syndic"
    ],
    "tgt": "minion01,minion02,master-syndic",
    "tgt_type": "list",
    "user": "saltdev"
}
minion/refresh/minion02 {
    "Minion data cache refresh": "minion02",
    "_stamp": "2018-06-06T14:01:12.137872"
}
syndic/master-syndic/minion/refresh/minion02    {
    "Minion data cache refresh": "minion02",
    "_stamp": "2018-06-06T14:01:12.148089"
}
minion/refresh/minion01 {
    "Minion data cache refresh": "minion01",
    "_stamp": "2018-06-06T14:01:12.149743"
}
syndic/master-syndic/minion/refresh/minion01    {
    "Minion data cache refresh": "minion01",
    "_stamp": "2018-06-06T14:01:12.153928"
}
salt/job/20180606140111583476/ret/minion02      {
    "_stamp": "2018-06-06T14:01:12.160757",
    "fun": "state.sls",
    "fun_args": null,
    "id": "minion02",
    "jid": "20180606140111583476",
    "out": "highstate",
    "retcode": 0,
    "return": {
        "cmd_|-testing_|-uptime_|-run": {
            "__id__": "testing",
            "__run_num__": 0,
            "__sls__": "test",
            "changes": {
                "pid": 5566,
                "retcode": 0,
                "stderr": "",
                "stdout": " 14:01:11 up  6:21,  0 users,  load average: 0.00, 0.00, 0.00"
            },
            "comment": "Command \"uptime\" run",
            "duration": 10.32,
            "name": "uptime",
            "result": true,
            "start_time": "14:01:11.929289"
        }
    },
    "success": true
}
minion/refresh/master-syndic    {
    "Minion data cache refresh": "master-syndic",
    "_stamp": "2018-06-06T14:01:12.169337"
}
salt/job/20180606140111583476/ret/minion01      {
    "_stamp": "2018-06-06T14:01:12.187343",
    "fun": "state.sls",
    "fun_args": null,
    "id": "minion01",
    "jid": "20180606140111583476",
    "out": "highstate",
    "retcode": 0,
    "return": {
        "cmd_|-testing_|-uptime_|-run": {
            "__id__": "testing",
            "__run_num__": 0,
            "__sls__": "test",
            "changes": {
                "pid": 5676,
                "retcode": 0,
                "stderr": "",
                "stdout": " 14:01:11 up  6:23,  0 users,  load average: 0.00, 0.02, 0.00"
            },
            "comment": "Command \"uptime\" run",
            "duration": 9.972,
            "name": "uptime",
            "result": true,
            "start_time": "14:01:11.892529"
        }
    },
    "success": true
}
salt/job/20180606140111583476/ret/master-syndic {
    "_stamp": "2018-06-06T14:01:12.305404",
    "cmd": "_return",
    "fun": "state.sls",
    "fun_args": [
        "test"
    ],
    "id": "master-syndic",
    "jid": "20180606140111583476",
    "out": "highstate",
    "retcode": 0,
    "return": {
        "cmd_|-test cmd_|-uptime_|-run": {
            "__id__": "test cmd",
            "__run_num__": 0,
            "__sls__": "test",
            "changes": {
                "pid": 24026,
                "retcode": 0,
                "stderr": "",
                "stdout": " 14:01:12 up  6:28,  1 user,  load average: 0.07, 0.03, 0.00"
            },
            "comment": "Command \"uptime\" run",
            "duration": 20.197,
            "name": "uptime",
            "result": true,
            "start_time": "14:01:12.172091"
        }
    },
    "success": true
}
```

salt event on syndic
```
20180606140111583476    {
    "_stamp": "2018-06-06T14:01:11.501716",
    "minions": [
        "minion01",
        "minion02"
    ]
}
salt/job/20180606140111583476/new       {
    "_stamp": "2018-06-06T14:01:11.506126",
    "arg": [
        "test"
    ],
    "fun": "state.sls",
    "jid": "20180606140111583476",
    "minions": [
        "minion01",
        "minion02"
    ],
    "missing": [
        "master-syndic"
    ],
    "tgt": "minion01,minion02,master-syndic",
    "tgt_type": "list",
    "user": "saltdev"
}
minion/refresh/minion02 {
    "Minion data cache refresh": "minion02",
    "_stamp": "2018-06-06T14:01:11.795207"
}
minion/refresh/minion01 {
    "Minion data cache refresh": "minion01",
    "_stamp": "2018-06-06T14:01:11.799555"
}
salt/job/20180606140111583476/ret/minion01      {
    "_stamp": "2018-06-06T14:01:11.933083",
    "cmd": "_return",
    "fun": "state.sls",
    "fun_args": [
        "test"
    ],
    "id": "minion01",
    "jid": "20180606140111583476",
    "out": "highstate",
    "retcode": 0,
    "return": {
        "cmd_|-testing_|-uptime_|-run": {
            "__id__": "testing",
            "__run_num__": 0,
            "__sls__": "test",
            "changes": {
                "pid": 5676,
                "retcode": 0,
                "stderr": "",
                "stdout": " 14:01:11 up  6:23,  0 users,  load average: 0.00, 0.02, 0.00"
            },
            "comment": "Command \"uptime\" run",
            "duration": 9.972,
            "name": "uptime",
            "result": true,
            "start_time": "14:01:11.892529"
        }
    },
    "success": true
}
salt/job/20180606140111583476/ret/minion02      {
    "_stamp": "2018-06-06T14:01:11.943244",
    "cmd": "_return",
    "fun": "state.sls",
    "fun_args": [
        "test"
    ],
    "id": "minion02",
    "jid": "20180606140111583476",
    "out": "highstate",
    "retcode": 0,
    "return": {
        "cmd_|-testing_|-uptime_|-run": {
            "__id__": "testing",
            "__run_num__": 0,
            "__sls__": "test",
            "changes": {
                "pid": 5566,
                "retcode": 0,
                "stderr": "",
                "stdout": " 14:01:11 up  6:21,  0 users,  load average: 0.00, 0.00, 0.00"
            },
            "comment": "Command \"uptime\" run",
            "duration": 10.32,
            "name": "uptime",
            "result": true,
            "start_time": "14:01:11.929289"
        }
    },
    "success": true
}
```

### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)

master conf of master of master
```
vagrant@master:~$ grep -v '^#' /etc/salt/master | grep -v '^$'
external_auth:
  pam:
    saltdev:
      - '*':
        - test.*
        - state.*
        - grains.*
        - pkg.*
rest_cherrypy:
  port: 8000
  host: 0.0.0.0
  disable_ssl: True
  log.access_file: /var/log/salt/salt_api.access.log
  log.error_file: /var/log/salt/salt_api.error.log
  debug: True
order_masters: True
```

master conf of syndic 
```
vagrant@master-syndic:~$ grep -v '^#' /etc/salt/master | grep -v '^$'
syndic_master:
  - master
```

### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)
use the above conf setup master, syndic, minion, the use the following script
```
vagrant@master:~$ cat cherry-t.sh
#!/bin/bash

set -x

curl http://localhost:8000/login \
    -c ./cookies.txt \
    -H 'Accept: application/x-yaml' \
    -d username=saltdev \
    -d password=saltdev \
    -d eauth=pam 2>/dev/null


curl http://localhost:8000 \
    -b ./cookies.txt \
    -H 'Accept: application/x-yaml' \
    -d client=local \
    -d tgt='minion01,minion02,master-syndic' \
    -d fun=state.sls \
    -d tgt_type=list \
    -d arg=test
```

put the following file on syndic
```
vagrant@master-syndic:~$ cat /srv/salt/test.sls
testing:
  cmd.run:
    - name: uptime
```

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)

```
vagrant@master:~$ salt --versions-report
Salt Version:
           Salt: 2018.3.0

Dependency Versions:
           cffi: Not Installed
       cherrypy: 3.5.0
       dateutil: 2.5.3
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.1
          ioflo: Not Installed
         Jinja2: 2.9.4
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, Jan 19 2017, 14:48:08)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.2.1

System Versions:
           dist: debian 9.2
         locale: UTF-8
        machine: x86_64
        release: 3.16.0-4-amd64
         system: Linux
        version: debian 9.2
```

**Comment by garethgreenaway** [comment]
@vinian Thanks for the report!

**Comment by vinian** [comment]
when i checkout the source code, found the problem was fixed in branch `2018.3`, but the source code(2018.3) in git repo was not match with the deb repo(2018.3.0+ds-1). 

deb package on my host
```
vagrant@master:~$ dpkg -l |grep salt
ii  salt-api                       2018.3.0+ds-1                  all          Generic, modular network access system
ii  salt-common                    2018.3.0+ds-1                  all          shared libraries that salt requires for all packages
ii  salt-master                    2018.3.0+ds-1                  all          remote manager to administer servers via salt
ii  salt-minion                    2018.3.0+ds-1                  all          client package for salt, the distributed remote execution system
```

here's my simple fix for the problem
```
vagrant@master:/usr/lib/python2.7/dist-packages/salt/client$ pwd
/usr/lib/python2.7/dist-packages/salt/client
vagrant@master:/usr/lib/python2.7/dist-packages/salt/client$ sudo git diff
diff --git a/__init__.py b/__init__.py
index cdc1ce4..99bea22 100644
--- a/__init__.py
+++ b/__init__.py
@@ -1330,8 +1330,8 @@ class LocalClient(object):
             for minion in list((minions - found)):
                 yield {minion: {'failed': True}}

-        if missing:
-            for minion in missing:
+        if set(missing) - found:
+            for minion in set(missing) - found:
                 yield {minion: {'failed': True}}

     def get_returns(
```

output of the test
```
vagrant@master:~$ bash cherry-t.sh
+ curl http://localhost:8000/login -c ./cookies.txt -H 'Accept: application/x-yaml' -d username=saltdev -d password=saltdev -d eauth=pam
return:
- eauth: pam
  expire: 1528405393.510493
  perms:
  - '*':
    - test.*
    - state.*
    - grains.*
    - pkg.*
  start: 1528362193.510492
  token: 3d608df9df87b69300d78e74d6ef9b745915bb78
  user: saltdev
+ curl http://localhost:8000 -b ./cookies.txt -H 'Accept: application/x-yaml' -d client=local -d tgt=minion01,minion02,master-syndic -d fun=test.ping -d tgt_type=list
return:
- master-syndic: true
  minion01: true
  minion02: true
```

```
vagrant@master:~$ bash cherry-t.sh
+ curl http://localhost:8000/login -c ./cookies.txt -H 'Accept: application/x-yaml' -d username=saltdev -d password=saltdev -d eauth=pam
return:
- eauth: pam
  expire: 1528405432.097989
  perms:
  - '*':
    - test.*
    - state.*
    - grains.*
    - pkg.*
  start: 1528362232.097989
  token: b8106a8022751138a75108b35e3bd54bc72f5354
  user: saltdev
+ curl http://localhost:8000 -b ./cookies.txt -H 'Accept: application/x-yaml' -d client=local -d tgt=minion01,minion02,master-syndic -d fun=state.sls -d arg=test -d tgt_type=list
return:
- master-syndic:
    cmd_|-test cmd_|-uptime_|-run:
      __id__: test cmd
      __run_num__: 0
      __sls__: test
      changes:
        pid: 17250
        retcode: 0
        stderr: ''
        stdout: ' 09:03:53 up 1 day,  1:31,  0 users,  load average: 0.00, 0.00, 0.00'
      comment: Command "uptime" run
      duration: 20.268
      name: uptime
      result: true
      start_time: '09:03:53.179470'
  minion01:
    cmd_|-testing_|-uptime_|-run:
      __id__: testing
      __run_num__: 0
      __sls__: test
      changes:
        pid: 6372
        retcode: 0
        stderr: ''
        stdout: ' 09:03:53 up 1 day,  1:26,  0 users,  load average: 0.00, 0.00, 0.00'
      comment: Command "uptime" run
      duration: 13.106
      name: uptime
      result: true
      start_time: '09:03:53.096773'
  minion02:
    cmd_|-testing_|-uptime_|-run:
      __id__: testing
      __run_num__: 0
      __sls__: test
      changes:
        pid: 6234
        retcode: 0
        stderr: ''
        stdout: ' 09:03:52 up 1 day,  1:24,  0 users,  load average: 0.00, 0.00, 0.00'
      comment: Command "uptime" run
      duration: 10.034
      name: uptime
      result: true
      start_time: '09:03:52.753170'
```

**Comment by terminalmage** [comment]
This is a duplicate of https://github.com/saltstack/salt/issues/47199, which was fixed by https://github.com/saltstack/salt/pull/47275 on April 24th.
